presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.bigquery;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.cloud.bigquery.Field;
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ColumnMetadata;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+public class BigQueryColumnHandle
+        implements ColumnHandle, BigQueryType.Adaptor
+{
+    private final String name;
+    private final BigQueryType bigQueryType;
+    private final Field.Mode mode;
+    private final List<BigQueryColumnHandle> subColumns;
+    private final String description;
+
+    @JsonCreator
+    public BigQueryColumnHandle(
+            @JsonProperty("name") String name,
+            @JsonProperty("bigQueryType") BigQueryType bigQueryType,
+            @JsonProperty("mode") Field.Mode mode,
+            @JsonProperty("subColumns") List<BigQueryColumnHandle> subColumns,
+            @JsonProperty("description") String description)
+    {
+        this.name = requireNonNull(name, "column name cannot be null");
+        this.bigQueryType = requireNonNull(bigQueryType, "column type cannot be null for column [" + name + "]");
+        this.mode = mode != null ? mode : Field.Mode.NULLABLE;
+        this.subColumns = subColumns != null ? subColumns : emptyList();
+        this.description = description;
+    }
+
+    @JsonProperty
+    public String getName()
+    {
+        return name;
+    }
+
+    @Override
+    @JsonProperty
+    public BigQueryType getBigQueryType()
+    {
+        return bigQueryType;
+    }
+
+    @Override
+    public ImmutableMap<String, BigQueryType.Adaptor> getBigQuerySubTypes()
```

Comment:
Hi @electrum , I agree that returning a base class rather than the implementation is a good idea, but in this case I'm not sure it applies. `ImmutableMap` is not really a `Map` implementation, as it lacks a lot of its functionality. If we go by the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle) then `Map` is a subclass of `ImmutableMap` as it provides all of `ImmutableMap` functionality, but the opposite is not true and can lead to an `UnsupportedOperationException`. What do you think?

Reply:
The map return here is for callers to read from. They have no reason to attempt to write to it, nor should they assume they can. Also, writing to a collection returned from a method that's not intended to be written to is bad practice (this can cause corruption or other bugs if the code was not written defensively in this regard).

We prefer to leave APIs flexible to allow changing the implementation. Using `Map` here allows the implementation to later change to `unmodifiableMap()`, an immutable map from JDK 9+, etc. This is similar to why we wouldn't declare or return `HashMap`.

You might also consider this a code style issue. I can understand the argument for using the `Immutable` interfaces everywhere possible in a code base, but we don't do that in Presto.